#15, #16 -  launch script

jillianderson8 opened the pull request:
#15 
#16

jillianderson8 committed:
#16 - changes to make launch script work on Windows and Unix

jillianderson8 committed:
#15 - disabling pip version check to prevent failures on pip warnings

jillianderson8 committed:
Updating to include some of the other improvements

jillianderson8 committed:
Adding in features implemented in #32

jillianderson8 committed:
adding powershell script, as added in PR #32

jillianderson8 commented:
@brad-griffin I've added in the additional changes you made in #32, but retained our basic structure, including argbash formatting. Although I cannot test certain features without being on a Windows device. Please let me know if you run into any trouble!

brad-griffin commented:
@jillianderson8 , I had a look through the script and it looked okay to me (although what do I really know about interacting with the operating system?!).

As a second check, I gave the script to claude and asked to check if there were any red flags. I got back the script in the htm below, with the following fixes. Most are more for aesthetics, but the system operator path delimiter is important, I think. Feel free to implement any or all of the changes.

[launch script.htm](https://github.com/user-attachments/files/24177157/launch.script.htm) (scroll to see comparison of script changes)

Key Fixes

- Cross-platform path splitting: Now uses ; on Windows and : on Unix-like systems (lines 201-205)
- Robust version sorting: Added a get_python_version() function and manual sorting fallback that works on all platforms, including macOS (lines 191-198, 272-302) 
  (the sort -rV command may not work correctly on all systems, particularly macOS, which doesn't support the -V flag for version sorting, causing the "recommended" Python selection to be incorrect)
- Windows Python detection: Fixed glob expansion by using unquoted variable expansion in a for loop (lines 231-241)
- Better version display: Shows actual Python version numbers (e.g., "Python 3.11.5") in the selection menu instead of just executable names
- Safe activation: Checks if the activation script exists before sourcing it (lines 335-349)
- Safe deactivation: Only calls deactivate if the function exists (lines 395-399)
- Better error handling: Added file existence checks for executables and activation scripts

brad-griffin commented:
One more question on this - does the script still work if someone has installed their version of Python through Anaconda? My reading on venv and conda sounds like they don't always play well together. Could we add a check like:

```
def running_under_conda():
    if os.environ.get("CONDA_PREFIX"):
        return True
    exe = sys.executable.lower()
    return "conda" in exe or "anaconda" in exe
```

And then, if they are using conda, activate their environment accordingly? We might need to grab the path for their parent environment directory since I think anaconda keeps everything centralised.

jillianderson8 committed:
Fix cross-platform compatibility and improve Python detection

- Fix path separator detection for Windows terminals (use ; or : depending on environment)
- Use OS detection consistently for virtual environment activation (Scripts/ vs bin/)
- Add get_python_version() function to display actual version numbers in selection menu
- Improve Python version sorting with manual fallback for systems without sort -V
- Add error handling for venv creation with Anaconda-specific guidance
- Add file existence checks for Python executables and activation scripts
- Add safe deactivation check to prevent errors if deactivate function doesn't exist
- Remove unused LBLUE color variable

jillianderson8 commented:
@brad-griffin Thanks for the thorough review and suggestions! I've implemented the key fixes you identified:

**Cross-platform path splitting** - Now correctly detects and uses `;` on Windows and `:` on Unix-like systems with proper regex escaping for Windows paths. The script actually checks which separator is in use since some Windows terminal environments (like Git Bash) use `:` like Unix systems.

**Virtual environment activation** - Uses OS detection consistently to check the correct directory (`Scripts/` on Windows, `bin/` on Unix)

**Safe deactivation** - Only calls deactivate if the function exists

---

Regarding your question about Anaconda:

Good catch. Many modern Anaconda installations actually work with `venv`, but some (particularly older versions) don't include the venv module by default. Rather than trying to detect Anaconda upfront and handle it specially, I've added error handling around the venv creation that will catch any Python installation that can't create a virtual environment:
```bash
print_color "Building ${_arg_venv_name}..." "$NC"
if ! "$SELECTED_PYTHON" -m venv "${_arg_venv_name}" 2>/dev/null; then
    print_color "FAILED\n" "$RED"
    print_color "The selected Python cannot create a virtual environment.\n" "$RED"
    
    # Check if this is Anaconda
    if "$SELECTED_PYTHON" -c "import sys; exit(0 if 'conda' in sys.version.lower() or 'anaconda' in sys.executable.lower() else 1)" 2>/dev/null; then
        print_color "This appears to be an Anaconda Python installation.\n" "$YLW"
        print_color "Some Anaconda installations don't include venv. Try:\n" "$YLW"
        print_color "  1. Using 'conda install venv' to add venv support\n" "$YLW"
        print_color "  2. Using 'conda create' instead for Anaconda environments\n" "$YLW"
        print_color "  3. Installing standard Python from python.org\n" "$YLW"
    else
        print_color "Try selecting a different Python installation.\n" "$YLW"
    fi
    exit 1
fi
print_color "DONE\n" "$LGRN"
```

This approach lets Anaconda Python work if it supports venv (which many do), provides helpful Anaconda-specific error messages if it doesn't, and keeps the script focused on standard venv rather than mixing conda environment management.

If users prefer conda, they can use `conda create` directly. Let me know if you'd prefer a different solution.

brad-griffin commented:
Works for me!
